pxf/pxf-jdbc/src/main/java/org/apache/hawq/pxf/plugins/jdbc/JdbcAccessor.java

```diff
@@ -0,0 +1,437 @@
+            );
+            batchSize = 0;
+        }
+
+        // Process poolSize
+        if (poolSize < 1) {
+            poolSize = Runtime.getRuntime().availableProcessors();
+            LOG.info(
+                "The POOL_SIZE is set to the number of CPUs available (" + Integer.toString(poolSize) + ")"
+            );
+        }
+        if (poolSize > 1) {
+            executorServiceWrite = Executors.newFixedThreadPool(poolSize);
+            poolTasks = new LinkedList<>();
+        }
+
+        // Setup WriterCallableFactory
+        writerCallableFactory = new WriterCallableFactory();
+        writerCallableFactory.setPlugin(this);
+        writerCallableFactory.setQuery(queryWrite);
+        writerCallableFactory.setBatchSize(batchSize);
+        if (poolSize == 1) {
+            writerCallableFactory.setStatement(statementWrite);
+        }
+
+        writerCallable = writerCallableFactory.get();
+
+        return true;
+    }
+
+	/**
+     * writeNextObject() implementation
+     *
+     * If batchSize is not 0 or 1, add a tuple to the batch of statementWrite
+     * Otherwise, execute an INSERT query immediately
+     *
+     * In both cases, a {@link java.sql.PreparedStatement} is used
+     *
+     * @throws SQLException if a database access error occurs
+     * @throws IOException if the data provided by {@link JdbcResolver} is corrupted
+     * @throws ClassNotFoundException if pooling is used and the JDBC driver was not found
+     * @throws IllegalStateException if writerCallableFactory was not properly initialized
+     * @throws Exception if it happens in writerCallable.call()
+     */
+    @Override
+    public boolean writeNextObject(OneRow row) throws Exception {
+        if (writerCallable == null) {
+            throw new IllegalStateException("The JDBC connection was not properly initialized (writerCallable is null)");
+        }
+
+        SQLException rollbackException = null;
+
+        writerCallable.supply(row);
+        if (writerCallable.isCallRequired()) {
+            if (poolSize > 1) {
+                // Pooling is used. Create new writerCallable
+                poolTasks.add(executorServiceWrite.submit(writerCallable));
+                writerCallable = writerCallableFactory.get();
+            }
+            else {
+                // Pooling is not used
+                try {
+                    rollbackException = writerCallable.call();
+                }
+                catch (SQLException e) {
+                    rollbackException = e;
+                }
+                catch (Exception e) {
+                    // This is not expected
+                    throw e;
+                }
+            }
+        }
+
+        if (rollbackException != null) {
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Rollback is now required");
+            }
+            rollbackException = tryRollback(statementWrite.getConnection(), rollbackException);

```

Comment:
we cannot guarantee transactional semantics for this profile anyways, as a single INSERT will be executed by multiple segments and multiple PXF threads, and there is no way for one of the threads to know the status of the others. It was suggested before to remove rollback logic altogether as it serves no practical purpose and just complicates the code and the processing logic. @leskin-in @kapustor -- What do you think ?

Reply:
@denalex , you are right. 
The only scenario when `tryRollback()` works properly is one PXF segment, one thread, and an external database that supports transactions.
There are more than one PXF segment available in all cases, thus there is no guarantee that rollback will work properly, and therefore, rollback is of no use.